kvserver: acquire token for admin scatter

wenyihu6 opened the pull request:
**kvserver: AdminScatter does not obtain an allocator token**

This commit adds a test, `TestAdminScatterAllocatorToken`, demonstrating
that `AdminScatter` issues replica changes without obtaining an
allocator token. This can lead to the replicate queue and scatter
racing with each other (and failing due to descriptor-changed errors).

Informs: #144579
Release note: None

Co-authored-by: Mira Radeva mira@cockroachlabs.com

---

**kvserver: refactor scatter operation out in adminScatter**

This commit moves the scatter operation logic of adminScatter into a separate
helper function to enhance clarity and simplify future commits.

Epic: none
Release note: none

---

**kvserver: logs suitable candidates for lease transfer during adminScatter**

This commit adds logging for suitable lease transfer candidates during
adminScatter, providing more visibility into ideal candidates.

Epic: none
Release note: none

---

**kvserver: acquire token for admin scatter**

Previously, we have seen descriptor changed errors during scatter due to
concurrent range descriptor changes triggered by the replicate queue. The
replicate queue encountered similar errors, suggesting both it and scatter may
have conflicted when modifying the same range.

To prevent contention like this, we previously introduced the concept of an
allocator token, acquired per-range by the leaseholder to coordinate replication
changes. It's already used by the store rebalancer and replicate queue to
synchronize changes to the same range.

This patch updates AdminScatter to acquire the allocator token before calling
processOneChange, aligning its behavior with replicateQueue.process. Note that
AdminScatter currently does not retry if the token isn't available. A future
commit will introduce a server side retry.

Note that we are unclear whether this is the root cause of failed scatter
operations since descriptor changed should be treated as a retriable error.

Resolves: #144579
Release note: none

---

**kvserver: retry if token acquisition failed**

Previously, adminScatter did not retry on token acquisition failures for replica
changes, relying on the client to retry. This commit introduces server-side
retries, attempting up to 5 times before returning to client.

Part of: #144579
Release note: none

miraradeva committed:
kvserver: AdminScatter does not obtain an allocator token

This commit adds a test, `TestAdminScatterAllocatorToken`, demonstrating
that `AdminScatter` issues replica changes without obtaining an
allocator token. This can lead to the replicate queue and scatter
racing with each other (and failing due to descriptor-changed errors).

Informs: #144579
Release note: None

Co-authored-by: Mira Radeva mira@cockroachlabs.com

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/144913)
<!-- Reviewable:end -->

wenyihu6 committed:
kvserver: refactor scatter operation out in adminScatter

This commit moves the scatter operation logic of adminScatter into a separate
helper function to enhance clarity and simplify future commits.

Epic: none
Release note: none

stevendanna reviewed (commented):
> Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144913)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OOXWPg5BZv6UkjbqoPz:-OOXWPg5BZv6UkjbqoQ-:b-83sjoh) (no related file):*
Nice!
___
*[`pkg/kv/kvserver/replica_command.go` line 4177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OOXW8jVD2Lsps47axLT:-OOXW8jVD2Lsps47axLU:bc0scir) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b95b74f68850785405950549212580c07a1a77e/pkg/kv/kvserver/replica_command.go#L4177)):*
> ```Go
> 	currentAttempt := 0
> 
> 	if tokenErr := r.allocatorToken.TryAcquire(ctx, rq.name); tokenErr != nil {
> ```

You have it in your commit message, but a comment reminding us that we need to take the token here because our use of processOneChange below bypasses where the replicate queue acquires the token.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144913)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvserver/replica_command.go` line 4177 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OOXW8jVD2Lsps47axLT:-OOXeMxiC1vlMtfALCKF:baf33ic) ([raw file](https://github.com/cockroachdb/cockroach/blob/1b95b74f68850785405950549212580c07a1a77e/pkg/kv/kvserver/replica_command.go#L4177)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

You have it in your commit message, but a comment reminding us that we need to take the token here because our use of processOneChange below bypasses where the replicate queue acquires the token.
</blockquote></details>

Good call, done. 


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Nice job getting this out so quickly! 

> Reviewed 2 of 2 files at r7, 1 of 1 files at r8, 2 of 2 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-:-OOZhQHXCfXYkz1jBseW:bd1bm8c)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @stevendanna and @wenyihu6)

___
*[`pkg/kv/kvserver/replica_command.go` line 4230 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OOZgtOBECgzUT2k_TL3:-OOZgtOBECgzUT2k_TL4:b-r1y3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc8374086a3fc2c7d83ec9dc4811b3b21ae276/pkg/kv/kvserver/replica_command.go#L4230)):*
> ```Go
> // AdminScatterResponse to tell clients to potentially retry the ScatterRequest
> // if ranges fail to move.
> func (r *Replica) adminScatter(
> ```

Would this all be simpler if we acquired the token once, up top in `adminScatter`, and returned a response with `NoReplicasMoved = true` if we failed to acquire the token? 

That then means we don't have to acquire the token twice, once for replicas and then again for leases. Is there any benefit to acquiring the token twice? 


<!-- Sent from Reviewable.io -->

wenyihu6 committed:
kvserver: acquire token for admin scatter

Previously, we have seen descriptor changed errors during scatter due to
concurrent range descriptor changes triggered by the replicate queue. The
replicate queue encountered similar errors, suggesting both it and scatter may
have conflicted when modifying the same range.

To prevent contention like this, we previously introduced the concept of an
allocator token, acquired per-range by the leaseholder to coordinate replication
changes. It's already used by the store rebalancer and replicate queue to
synchronize changes to the same range.

This patch updates AdminScatter to acquire the allocator token before calling
processOneChange, aligning its behavior with replicateQueue.process. Note that
AdminScatter currently does not retry if the token isn't available. A future
commit will introduce a server side retry.

Note that we are unclear whether this is the root cause of failed scatter
operations since descriptor changed should be treated as a retriable error.

Resolves: #144579
Release note: none

wenyihu6 committed:
kvserver: logs suitable candidates for lease transfer during adminScatter

This commit adds logging for suitable lease transfer candidates during
adminScatter, providing more visibility into ideal candidates.

Epic: none
Release note: none

wenyihu6 committed:
kvserver: retry if token acquisition failed

Previously, adminScatter did not retry on token acquisition failures for replica
changes, relying on the client to retry. This commit introduces server-side
retries, attempting up to 5 times before returning to client.

Part of: #144579
Release note: none

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144913)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @stevendanna)

___
*[`pkg/kv/kvserver/replica_command.go` line 4230 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OOZgtOBECgzUT2k_TL3:-OOen8Qa6Ho4GzFWvIl1:b-1imeql) ([raw file](https://github.com/cockroachdb/cockroach/blob/27cc8374086a3fc2c7d83ec9dc4811b3b21ae276/pkg/kv/kvserver/replica_command.go#L4230)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Would this all be simpler if we acquired the token once, up top in `adminScatter`, and returned a response with `NoReplicasMoved = true` if we failed to acquire the token? 

That then means we don't have to acquire the token twice, once for replicas and then again for leases. Is there any benefit to acquiring the token twice?
</blockquote></details>

Good call, done. Don't see any harms doing it. Also added the retry backoff we discussed offline in the last commit.


<!-- Sent from Reviewable.io -->

kvoli reviewed (commented):
> Reviewed 2 of 2 files at r10, 1 of 1 files at r11, 1 of 1 files at r12, 2 of 2 files at r13, 1 of 1 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144913)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @wenyihu6)

___
*[`-- commits` line 5 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9oIpy67OYcVdzuoG_:-OQ9oIpy67OYcVdzuoGa:b6x26jt):*
nit: AdminScatter
___
*[`-- commits` line 48 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9rgQ3DRW1OB_HDK8p:-OQ9rgQ3DRW1OB_HDK8q:bey892l):*
In the follow up commit here, a retry loop on the server side is introduced to retry when the token is unavailable. Is the intention to return the error to the client or retry server side?



_Code quote:_
```
  replicate queue, we do not retry if the token isn't available and rely on client
  side to retry.
```
___
*[`pkg/kv/kvserver/replica_command.go` line 4239 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9qSNV78T5WALyWlKn:-OQ9qSNV78T5WALyWlKo:b-9c9lcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5e79fde7871b67340444a733731a2b5c08f2a5/pkg/kv/kvserver/replica_command.go#L4239)):*
> ```Go
> 					log.VEventf(ctx, 2, "randomly transferring lease to s%d", targetStoreID)
> 					if err := r.AdminTransferLease(ctx, targetStoreID, false /* bypassSafetyChecks */); err != nil {
> 						log.Warningf(ctx, "scatter lease to s%d failed due to %+v: candidates included %v",
> ```

This was already here but might as well update the error to just be a `%v`, it should format.
___
*[`pkg/kv/kvserver/replica_command.go` line 4262 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9q7t8DGKSgCDvCwLe:-OQ9q7t8DGKSgCDvCwLf:b7yninf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d64824d565bc0ca295dc6455ff5fac37a7b0039f/pkg/kv/kvserver/replica_command.go#L4262)):*
> ```Go
> // best-effort. Ranges that cannot be moved will just return early and not
> // return an error. NoReplicasMoved is included in AdminScatterResponse to tell
> // clients to potentially retry the ScatterRequest if ranges fail to move.
> ```

The latter part of this comment appears stale with https://github.com/cockroachdb/cockroach/pull/146129.



_Code quote:_
```Go
// return an error. NoReplicasMoved is included in AdminScatterResponse to tell
// clients to potentially retry the ScatterRequest if ranges fail to move.
```


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144913)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @kvoli, and @stevendanna)

___
*[`-- commits` line 5 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9oIpy67OYcVdzuoG_:-OQATdjcCHTKKdoBk_PW:b-896fix):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: AdminScatter
</blockquote></details>

Done.
___
*[`-- commits` line 48 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9rgQ3DRW1OB_HDK8p:-OQATjr46XSzPAEzXrDt:b2v7zqt):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

In the follow up commit here, a retry loop on the server side is introduced to retry when the token is unavailable. Is the intention to return the error to the client or retry server side?
</blockquote></details>

This comment was stale. Originally we planned to rely on client retries but ended up doing some retries on the server side. Updated the commit messages. 
___
*[`pkg/kv/kvserver/replica_command.go` line 4262 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9q7t8DGKSgCDvCwLe:-OQATyrG6Vyxs0jYwbd5:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d64824d565bc0ca295dc6455ff5fac37a7b0039f/pkg/kv/kvserver/replica_command.go#L4262)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

The latter part of this comment appears stale with https://github.com/cockroachdb/cockroach/pull/146129.
</blockquote></details>

Removed.
___
*[`pkg/kv/kvserver/replica_command.go` line 4239 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-OQ9qSNV78T5WALyWlKn:-OQATwyw1f5yH54IfVxE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5e79fde7871b67340444a733731a2b5c08f2a5/pkg/kv/kvserver/replica_command.go#L4239)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

This was already here but might as well update the error to just be a `%v`, it should format.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r15, 1 of 1 files at r16, 1 of 1 files at r17, 2 of 2 files at r18, 1 of 1 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/144913#-:-OQE7tpn5yLaKshYFXMb:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @wenyihu6)



<!-- Sent from Reviewable.io -->

wenyihu6 commented:
TFTR!

bors r=kvoli

wenyihu6 commented:
blathers backport-all
